### ISSUE TITLE
[Memory Leak] GPU Memory Accumulates After Each Epoch in Training

### ISSUE THREAD

**Issue by JiruWang19971997** [OpenRLHF/OpenRLHF#1086]
<img width="524" height="304" alt="Image" src="https://github.com/user-attachments/assets/38f2d66e-bc0d-464d-9a5b-e5348e8112af" />

Hi, I'm using the latest commit of OpenRLHF with the following training script:
`set -x 

export PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True

ray job submit --address="http://127.0.0.1:8267" \
   --runtime-env-json='{"working_dir": "/root/OpenRLHF"}' \
   -- python3 -m openrlhf.cli.train_ppo_ray \
    --ref_num_nodes 1 \
    --ref_num_gpus_per_node 1 \
    --reward_num_nodes 0 \
    --reward_num_gpus_per_node 0 \
    --critic_num_nodes 1 \
    --critic_num_gpus_per_node 1 \
    --actor_num_nodes 1 \
    --actor_num_gpus_per_node 1 \
    --vllm_num_engines 1 \
    --vllm_tensor_parallel_size 1 \
    --pretrain /root/autodl-fs/qwen \
    --remote_rm_url /root/OpenRLHF/openrlhf/trainer/ppo_utils/hard_reward_label.py \
    --critic_pretrain /root/autodl-fs/qwen \
    --eval_n_samples_per_prompt 1 \
    --save_path /root/autodl-tmp/save \
    --micro_train_batch_size 4 \
    --train_batch_size 32 \
    --micro_rollout_batch_size 32 \
    --rollout_batch_size 320 \
    --max_epochs 1 \
    --num_episodes 10 \
    --label_key answer \
    --prompt_max_len 1024 \
    --generate_max_len 1024 \
    --zero_stage 3 \
    --bf16 \
    --eval_steps 1 \
    --actor_learning_rate 5e-7 \
    --critic_learning_rate 9e-6 \
    --init_kl_coef 0.01 \
    --prompt_data /root/OpenRLHF/data/train.json \
     --eval_dataset /root/OpenRLHF/data/test.json \
    --input_key input \
    --apply_chat_template \
    --max_samples 10000000 \
    --packing_samples \
    --normalize_reward \
    --adam_offload \
    --flash_attn \
    --vllm_sync_backend nccl \
    --gradient_checkpointing \
    --use_wandb b718e83ae2e0c39c7d64d7ffc96cae4c4a8253c6
      #   \
`
and observe the GPU memory increase after each epoch, and will cause a OOM after 10 epoches.

**Comment by hijkzzz** [comment]
maybe due to responses length become longer? please try to use dynamic batch
